### PR TITLE
Downstream merge of scream/eamxx fork into E3SM

### PR DESCRIPTION
**Pull request by jgfouca**
Downstream merge of scream/eamxx fork into E3SM
    
The build system fixes on the fork are needed for E3SM. 
Also updates E3SM with latest version of eamxx.
    
Merge remote-tracking branch 'scream/master' into jgfouca/scream_downstream_merge

Updates submodules:
* ekat
* haero
* mam4xx

EAM updates:
* Consolidate shoc_Ck nml settings
* New iop_coriolis nml setting (for nudging)
* initialize IOP information on the first restart time step
* SHOC updates
* Add wet/dry qv_sat (saturation mixing ratio), old qv_sat is still available for backwards compatibility
* P3 updates (scream version only)

This PR now includes https://github.com/E3SM-Project/E3SM/pull/5925 :

Update the YAKL submodule, and update PAM submodule to work with the YAKL update.

[YAKL]
[BFB] - (except for SCREAM/EAMXX and MMF2 cases)

### CODE REVIEW

**Comment by jgfouca** [comment]
Conflicts were trivial and easy to resolve.

**Comment by ndkeen** [comment]
I'm trying the branch.  The one GPU test listed in tests.py *does* build and work on pm-gpu -- so that's great.

However, I went ahead and launched some tests on pm-cpu and I'm seeing build fails.

```
/global/cfs/cdirs/e3sm/ndk/repos/jgfouca_scream_downstream_merge/components/eam/src/physics/p3/eam/micro_p3.F90(71): error #6580: Name in only-list does not exist or is not accessible.   [QV_SAT]
   use wv_sat_scream, only:qv_sat
---------------------------^
/global/cfs/cdirs/e3sm/ndk/repos/jgfouca_scream_downstream_merge/components/eam/src/physics/p3/eam/micro_p3.F90(480): error #6404: This name does not have a type, and must have an explicit type.   [QV_SAT]
       qv_sat_l(k)     = qv_sat(t_atm(k),pres(k),0)
-------------------------^
```

`ERP_Ld3.ne30pg2_EC30to60E2r2.WCYCL1850.pm-cpu_intel.allactive-pioroot1`

**Comment by jgfouca** [comment]
OK, i will try that case.

**Comment by jgfouca** [comment]
@ndkeen , I pushed a fix. I'm fairly confident that it's the right fix but I'm not 100% sure. I hope someone can double check it. The fix will at least fix the build problem for sure.

**Comment by oksanaguba** [comment]
Jim's fix is probably due to wetdry PR and changes https://github.com/E3SM-Project/scream/pull/2193/files#diff-c2a61193569e3531bfc444609ede48051f9327c27181948890ebabbaa152c9a4 .

**Comment by jgfouca** [comment]
@oksanaguba , yep. The eam micro_p3 used qv_sat in a couple additional places compared to eamxx micro_p3 (in a subroutine called `prevent_ice_overdepletion` which doesn't exist on the eamxx side). I assumed those usages should also be qv_sat_dry.

**Comment by rljacob** [comment]
@ndkeen was the one test you tried this one: SMS_Ln9.ne4pg2_ne4pg2.F2010-MMF1 from the e3sm_gpucxx suite?  If so, please try this one that @brhillman was trying to add in #5495 ERP_Ln9.ne4pg2_ne4pg2.F2010-SCREAMv1.

**Comment by jgfouca** [comment]
@rljacob , I verified these two cases work:

```
PASS SMS_Ln9.ne4pg2_ne4pg2.F2010-MMF1.mappy_gnu9 MODEL_BUILD
    Case dir: /ascldap/users/jgfouca/acme/scratch/SMS_Ln9.ne4pg2_ne4pg2.F2010-MMF1.mappy_gnu9.20240118_102217_nqa99n
PASS SMS_Ln9.ne4pg2_ne4pg2.F2010-SCREAMv1.mappy_gnu9 MODEL_BUILD
    Case dir: /ascldap/users/jgfouca/acme/scratch/SMS_Ln9.ne4pg2_ne4pg2.F2010-SCREAMv1.mappy_gnu9.20240118_102217_nqa99n
test-scheduler took 202.1448359489441 seconds
```

@ndkeen told me to try ERP_Ld3.ne30pg2_EC30to60E2r2.WCYCL1850, which did have a problem, which I have since fixed.

**Comment by rljacob** [comment]
That's good to know what both MMF and SCREAM can now build with the same gpu-cxx compiler settings!

File: .github/workflows/gh-pages.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# This workflow aims to automatically rebuild eamxx documentation
```
Comment (rljacob) [inline .github/workflows/gh-pages.yml:1]: I think this should be removed.  Or it needs to be blocked out with something like "if: ${{ github.event.repository.name = 'scream' }}"

Comment (bartgol) [reply .github/workflows/gh-pages.yml:1]: Good catch!

File: cime_config/machines/cmake_macros/crayclang-scream_frontier-scream-gpu.cmake [hunk 2]
```diff
@@ -0,0 +1,38 @@
+set(MPICC "mpicc")
```
Comment (rljacob) [inline cime_config/machines/cmake_macros/crayclang-scream_frontier-scream-gpu.cmake:1]: I thought the merge of build systems was going to get rid of all these crayclang-scream cmake files.

File: cime_config/machines/config_machines.xml [hunk 3]
```diff
@@ -1207,6 +1211,92 @@
     </environment_variables>
   </machine>
 
+  <machine MACH="frontier-scream-gpu">
```
Comment (rljacob) [inline cime_config/machines/config_machines.xml:1214]: same comment as above:  we shouldn't need scream-specific mach defines.

File: components/eamxx/CMakeLists.txt [hunk 4]
```diff
@@ -66,18 +67,19 @@ if (NOT SCREAM_CIME_BUILD)
...
-    OUTPUT_VARIABLE LAST_GIT_COMMIT_SHA
-    OUTPUT_STRIP_TRAILING_WHITESPACE)
-  set(LAST_GIT_COMMIT_SHA ${LAST_GIT_COMMIT_SHA} CACHE STRING "The sha of the last git commit.")
-  message(STATUS "The sha of the last commit is ${LAST_GIT_COMMIT_SHA}")
 else()
   # Ensure our languages are all enabled
   enable_language(C CXX Fortran)
 endif()
 
+# Print the sha of the last commit (useful to double check which version was tested on CDash)
```
Comment (bartgol) [inline components/eamxx/CMakeLists.txt:75]: I don't recall why we moved this out of the if statement above. Not that it matters much, but maybe sha info should be printed at the top of E3SM cmake machinery (easy to find).

**Comment by jgfouca** [comment]
@rljacob , the scream-specific stuff needs to remain for now. We can look at further unification once the repos are unified.

**Comment by rljacob** [comment]
This will need to be BFB if it is to go in before the maint-3.0 branch is made.

**Comment by jgfouca** [comment]
@rljacob , is there a specific test I should check to see if it's BFB? @mahf708 had the same concern that this may not be BFB.

**Review by bartgol — APPROVED** [review]

**Comment by rljacob** [comment]
Try SMS_Ln5.ne4pg2_oQU480.F2010.mappy_gnu

File: components/eam/src/physics/p3/eam/micro_p3_utils.F90 [hunk 5]
```diff
@@ -31,7 +31,7 @@ module micro_p3_utils
     real(rtype), public, parameter :: piov6 = pi*sxth
 
     ! maximum total ice concentration (sum of all categories)
-    real(rtype), public, parameter :: max_total_ni = 500.e+3_rtype  ! (m)
+    real(rtype), public, parameter :: max_total_ni = 740.e+3_rtype  ! (m)
```
Comment (rljacob) [inline components/eam/src/physics/p3/eam/micro_p3_utils.F90:34]: This looks like it will change answers.

**Comment by jgfouca** [comment]
I've confirmed the following tests from e3sm_developer DIFF on this branch:

```
DIFF ERP_Ln18.ne4_oQU240.F2010.mappy_gnu (phase BASELINE)
DIFF ERS_D.ne4_oQU240.F2010.mappy_gnu.eam-hommexx (phase BASELINE)
DIFF ERS_Ld3.ne4pg2_oQU480.F2010.mappy_gnu.eam-thetahy_sl_pg2 (phase BASELINE)
DIFF ERS_Ld3.ne4pg2_oQU480.F2010.mappy_gnu.eam-thetahy_sl_pg2_ftype0 (phase BASELINE)
DIFF SMS.ne4_oQU240.F2010.mappy_gnu.eam-cosplite (phase BASELINE)
DIFF SMS_Ln5.ne4pg2_oQU480.F2010.mappy_gnu (phase BASELINE)
DIFF SMS_Ln5.ne4pg2_oQU480.F2010.mappy_gnu.eam-thetahy_pg2 (phase BASELINE)
DIFF SMS_Ln5.ne4pg2_oQU480.F2010.mappy_gnu.eam-thetahy_sl_pg2 (phase BASELINE)
DIFF SMS_Ln5.ne4pg2_oQU480.F2010.mappy_gnu.eam-thetahy_sl_pg2_ftype0 (phase BASELINE)
DIFF SMS_Ln9.ne4_oQU240.F2010.mappy_gnu.eam-outfrq9s (phase BASELINE)
DIFF SMS_Ln9_P24x1.ne4_ne4.FDPSCREAM-ARM97.mappy_gnu (phase BASELINE)
DIFF SMS_P12x2.ne4_oQU240.WCYCL1850NS.mappy_gnu.allactive-mach_mods (phase BASELINE)
DIFF SMS_R_Ld5.ne4_ne4.FSCM-ARM97.mappy_gnu.eam-scm (phase BASELINE)
```

I'm going to re-run with the max_total_ni change reverted.

**Review by ndkeen — APPROVED** [review]
after fix, tests worked for me on pm-cpu

**Comment by jgfouca** [comment]
After a couple minor changes:
1. Reverting the one line change in `eam/src/physics/p3/eam/micro_p3_utils.F90` that changed a constant
2. Reverting the implementation change of qv_sat (renamed to qv_sat_dry) , specifically this line:
```
-    qv_sat_dry = ep_2*e_pres/max(1.e-3_rtype,p_atm_dry)
+    qv_sat_dry = ep_2*e_pres/max(1.e-3_rtype,(p_atm-e_pres))
```

All but one DIFF in e3sm_developer goes away. The remaining DIFF is in `SMS_Ln9_P24x1.ne4_ne4.FDPSCREAM-ARM97.mappy_gnu` and I will see if I can pinpoint what is causing that DIFF.

**Comment by jgfouca** [comment]
Oh, I should have realized that `SMS_Ln9_P24x1.ne4_ne4.FDPSCREAM-ARM97.mappy_gnu` is a scream atm case, so that DIFF is expected.

**Comment by rljacob** [comment]
Doc build warnings (which prevent build because of strict mode):
```
WARNING -  A relative path to 'the-c-e3sm-atmosphere-model-eamxx/common/eamxx_params.md' is included in the 'nav' configuration, which is not found in the documentation files.
WARNING -  Doc file 'the-c-e3sm-atmosphere-model-eamxx/user/model_input.md' contains a relative link '../common/eamxx_params.md', but the target 'the-c-e3sm-atmosphere-model-eamxx/common/eamxx_params.md' is not found among documentation files.
```

**Comment by bartgol** [comment]
> Doc build warnings (which prevent build because of strict mode):
> 
> ```
> WARNING -  A relative path to 'the-c-e3sm-atmosphere-model-eamxx/common/eamxx_params.md' is included in the 'nav' configuration, which is not found in the documentation files.
> WARNING -  Doc file 'the-c-e3sm-atmosphere-model-eamxx/user/model_input.md' contains a relative link '../common/eamxx_params.md', but the target 'the-c-e3sm-atmosphere-model-eamxx/common/eamxx_params.md' is not found among documentation files.
> ```

In the SCREAM repo, that file is autogenerated during the gh-pages action. There's a script in `eamxx/scripts` that is run by the gh workflow, which generates `eamxx_params.md` automatically from the XML defaults. Maybe we should isolate that action in its own action yaml file (something like `eamxx_autogen_params_docs.yaml`), and have e3sm run that script as part of the `e3sm-gh-pages.yaml` action. @mahf708 do you know off the top of your head how to achieve this? Otherwise I can look into this next week.

File: .github/workflows/eamxx-gh-pages.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  workflow_dispatch:
+
+concurrency:
+  # Prevent 2+ copies of this workflow from running concurrently
+  group: eamxx-docs-action
+
+jobs:
+
+  eamxx-docs:
+    if: ${{ github.event.label.name == 'EAMxx' }}
```
[the remark this answers is not in the document]
Comment (bartgol) [reply .github/workflows/eamxx-gh-pages.yml:37]: Agreed. The repo name cannot be set by mistake.

[the remark this answers is not in the document]
Comment (jgfouca) [reply .github/workflows/eamxx-gh-pages.yml:37]: But the scream repository is hopefully going away soon.

[the remark this answers is not in the document]
Comment (jgfouca) [reply .github/workflows/eamxx-gh-pages.yml]: @mahf708 , great, thanks for clearing that up.

**Comment by jgfouca** [comment]
Merged to next.

**Comment by jgfouca** [comment]
@rljacob , done

**Comment by rljacob** [comment]
There should also be a description of what's changing in EAM.

**Comment by rljacob** [comment]
Since you're updating ekat, how does this PR relate to #6101

**Comment by jgfouca** [comment]
@rljacob , I added an EAM change list. There might be a minor conflict with that other EKAT PR, should be very easy to resolve (just take the newer sha).

**Comment by bartgol** [comment]
> Since you're updating ekat, how does this PR relate to #6101

That PR is relatively orthogonal. The only thing it does is updating the kokkos submodule in ekat (and adjusting for API changes). I am trying to get all of that testing completed. but it's taking me a long time. I don't want that PR to block this one in any way. I will adjust that one if need be.

**Review by rljacob — APPROVED** [review]

**Comment by rljacob** [comment]
2 SCREAM diffs, as expected, but 1 MMF build fail:  SMS_Ln3.ne4pg2_ne4pg2.F2010-MMF2

**Comment by ambrad** [comment]
I think this PR is causing these new build failures: https://my.cdash.org/build/2477124

**Comment by jgfouca** [comment]
I think so too; I'm looking at the three cases on chrysalis now.

**Comment by jgfouca** [comment]
I've added https://github.com/E3SM-Project/E3SM/pull/5925 to this PR since both branches depend on each other now.

**Comment by rljacob** [comment]
@jgfouca be sure to update the description then.

**Pull request merged**